File: src/test/java/com/example/selenium/spree/pages/ProductPage.java
package com.example.selenium.spree.pages;

import org.junit.Assert;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.internal.Coordinates;
import org.openqa.selenium.internal.Locatable;
import org.openqa.selenium.remote.RemoteWebDriver;
import org.openqa.selenium.remote.RemoteWebElement;
import org.openqa.selenium.support.CacheLookup;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.ui.ExpectedCondition;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class ProductPage extends SpreePage {

  @CacheLookup
  @FindBy(css="#main-image>img")
  private WebElement mainImg;;

  @FindBy(name="quantity")
  private WebElement quantityTF;

  @FindBy(tagName="button")
  private WebElement addToCartBtn;
  
  private static String buildUrl(String productName) {
    return "http://spree.newcircle.com/products/" + 
        productName.replace(" ", "-").replace(".", "").toLowerCase();
  }
  
  public static ProductPage open(RemoteWebDriver webDriver, String productName) {
    webDriver.get(buildUrl(productName));
  
    return new ProductPage(webDriver, productName);
  }

  
  public ProductPage(RemoteWebDriver webDriver, String productName) {
    super(webDriver, buildUrl(productName));
  }

  public void validateTitle() {
    Assert.assertEquals("Spree Tote - Spree Demo Site", webDriver.getTitle());
  }

  public void validateImageSrc(String expected) {
    String actual = mainImg.getAttribute("src");
    String msg = "Found " + actual;
    Assert.assertTrue(msg, actual.startsWith(expected));
  }

  public void mouseOverThumbnail(int i) throws Exception {
    //webDriver.getMouse().mouseDown(where);
    
    i += 1;
    String path = ".//*[@id='product-thumbnails']/li["+i+"]";
    RemoteWebElement thumbnail = (RemoteWebElement)webDriver.findElementByXPath(path);
    // Coordinates co = ((Locatable)thumbnail).getCoordinates();
    webDriver.getMouse().mouseMove(thumbnail.getCoordinates());

    System.out.println("After we move the mouse");
    Thread.sleep(10*1000);
}

  public void clickThumbnail(int i) {
    i += 1;
    String path = ".//*[@id='product-thumbnails']/li["+i+"]/a";
    WebElement thumbnail = webDriver.findElementByXPath(path);
    thumbnail.click();
  }

  public void setQuantity(int i) {
    quantityTF.clear();
    quantityTF.sendKeys("3");
  }

  public CartPage clickAddToCart() {
    addToCartBtn.click();

    ExpectedCondition<Boolean> expectation = ExpectedConditions
        .urlToBe("http://spree.newcircle.com/cart");
    new WebDriverWait(webDriver, 5).until(expectation);

    return new CartPage(webDriver);
  }
}

File: python-course/com/example/selenium/spree/product_page.py
from selenium.webdriver.common.by import By

from com.example.selenium.spree.spree_page import SpreePage, url_contains, url_to_be


class ProductPage(SpreePage):
    
    def __init__(self, test, product:str):
        super().__init__(test)

        product = product.replace(" ", "-")
        product = product.lower()

        from selenium.webdriver.support.wait import WebDriverWait
        WebDriverWait(self.test.webDriver, 30).until(
            url_to_be("https://spreecommerce-demo.herokuapp.com/products/" + product)
        )

    def validateTitle(self):
        title = self.webDriver.title
        self.test.assertEquals("Spree Tote - Spree Demo Site", title)

    def validateUrl(self):
        url = self.webDriver.current_url
        self.test.assertEquals("https://spreecommerce-demo.herokuapp.com/products/spree-tote", url)

    def validateImageSrc(self, expected_url):
        img = self.webDriver.find_element_by_css_selector(".panel-body.text-center>img")
        actual_url = img.get_attribute("src")
        
        msg = "Found " + actual_url
        self.test.assertTrue(expected_url in actual_url, msg)
        
    def clickThumbnail(self, index):
        thumb = self.webDriver.find_element_by_xpath(".//*[@id='product-thumbnails']/li[" + str(index+1) + "]/a/img")
        thumb.click()
        
    def setQuantity(self, count):
        quantity_tf = self.webDriver.find_element_by_name("quantity")
        quantity_tf.clear()
        quantity_tf.send_keys( str(count) )
        
    def clickAddToCart(self):
        addBtn = self.webDriver.find_element_by_tag_name("button")
        addBtn.click()
        
    def confirmCartResponse(self, expected):

        by = (By.CLASS_NAME, 'alert-success')
        
        from selenium.webdriver.support.wait import WebDriverWait
        from selenium.webdriver.support.expected_conditions import text_to_be_present_in_element
        WebDriverWait(self.test.webDriver, 5).until(
            text_to_be_present_in_element(by, expected)
        )

        # response = self.webDriver.find_element_by_class_name("alert-success")
        # self.test.assertEquals(expected, response.text)
